src/providers/postgres/qgspostgresprovider.cpp

```diff
@@ -2029,15 +2029,41 @@ bool QgsPostgresProvider::addFeatures( QgsFeatureList &flist, Flags flags )
       delim = ',';
     }
 
-    if ( mPrimaryKeyType == PktInt || mPrimaryKeyType == PktFidMap || mPrimaryKeyType == PktUint64 )
+    // Optimization: if we have a single primary key column whose default value
+    // is a sequence, and that none of the features have a value set for that
+    // column, then we can completely omit inserting it.
+    bool skipSinglePKField = false;
+
+    if ( ( mPrimaryKeyType == PktInt || mPrimaryKeyType == PktFidMap || mPrimaryKeyType == PktUint64 ) )
     {
-      Q_FOREACH ( int idx, mPrimaryKeyAttrs )
+      if ( mPrimaryKeyAttrs.size() == 1 &&
+           defaultValueClause( mPrimaryKeyAttrs[0] ).startsWith( "nextval(" ) )
       {
-        insert += delim + quotedIdentifier( field( idx ).name() );
-        values += delim + QStringLiteral( "$%1" ).arg( defaultValues.size() + offset );
-        delim = ',';
-        fieldId << idx;
-        defaultValues << defaultValueClause( idx );
+        bool foundNonNullPK = false;
+        int idx = mPrimaryKeyAttrs[0];
+        for ( int i = 0; i < flist.size(); i++ )
+        {
+          QgsAttributes attrs2 = flist[i].attributes();
+          QVariant v2 = attrs2.value( idx, QVariant( QVariant::Int ) );
+          if ( !v2.isNull() )
+          {
+            foundNonNullPK = true;
+            break;
+          }
+        }
+        skipSinglePKField = !foundNonNullPK;
+      }
+
+      if ( !skipSinglePKField )
+      {
+        Q_FOREACH ( int idx, mPrimaryKeyAttrs )
```

Comment:
range based for instead of Q_FOREACH

Reply:
ok, note that I didn't modify that code. Was just put inside a if ()